### PR TITLE
[DRAFT][Blocked] Mem efficient attention - FW pass

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
First take for #161, only the forward pass with this PR (no training possible). The method is described [here](https://arxiv.org/pdf/2112.05682v2.pdf), the gist is that you compute the attention with you current best guess for the softmax renormalization, save your offset, and correct post-hoc 

TODO: 
- [ ] check parity all along (buggy for now)
- [ ]  possibly change the scheduling to improve L2 rate (may not be a limiting factor really)
- [ ] handle bias
- [ ] change the way we handle batch dimension ?
- [ ] normalize in the kernel if there's only ever one tile over N
- [ ] tiling case is broken by a factor of kN
cc @ptillet

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: xformers/triton/k_mem_efficient_attention.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    # We proceed per chunk over L, and save as we go
+    i = 0
+    rl = tl.arange(0, BLOCK_L)
+    for _ in range(L, 0, -BLOCK_L):
+        rl_i = rl + i * BLOCK_L
+
+        v_ptrs = V + rn[:, None] * L + rl_i[None, :]    # (BLOCK_N, BLOCK_L)
+        v = tl.load(v_ptrs, mask=((rn[:, None] < N) & (rl_i[None, :] < L)), other=0.0)
+
+        qkv = tl.dot(exp_acc, v).to(tl.float32)         # (BLOCK_M, BLOCK_L)
```
Comment (blefaudeux) [inline xformers/triton/k_mem_efficient_attention.py:78]: @ptillet not having .to(tl.float32) means that this crashes for instance, it's not really obvious to me why

**Comment by blefaudeux** [comment]
looks like the final buffer write breaks, but it's not completely obvious to me why that is. Visualizing the errors it looks like a stride mismatch, but not constant, maybe a race condition on top of that, a bit puzzling

Screenshot from 2021-12-21 21-41-45
Screenshot from 2021-12-21 20-43-21
to repro `CUDA_LAUNCH_BLOCKING=1 pytest tests -k efficient_attention -x -v`

**Comment by blefaudeux** [comment]
quick update: working around a bug in the Triton compiler, PoC is there and runs, not shippable as is. Lots of perf potential, the FW could actually be faster than a vanilla take while using a lot less memory. The BW will always be a little slower but the end result could well be worth it

**Comment by ptillet** [comment]
Yep there's definitely a bug in the compiler. The holiday break seems like the right time to rewrite how Triton handles data layout. I think there's a lot of potential for this kind of fused attention function to be very performant and memory-efficient.

**Comment by blefaudeux** [comment]
keeping the branch up but closing the PR, I cannot do much on this topic at the moment, dependent on upstream fixes on Triton

File: xformers/triton/k_mem_efficient_attention.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        q_, k_, v_ = q, k, v
+
+    B, M, L = q_.shape
+    B, N, L = k_.shape
+
+    BLOCK_M = 8
+    BLOCK_N = min(triton.next_power_of_2(N), 1024)  # increase the ceiling to save more memory
+    BLOCK_L = 8
+
+    tiles_n = triton.cdiv(N, BLOCK_N)
```
Comment (blefaudeux) [inline xformers/triton/k_mem_efficient_attention.py:133]: cc @dianaml0

**Comment by ptillet** [comment]
Hey! FYI mem-efficient attention is becoming a bigger priority for us. The bug is pretty deep inside of the compiler but I am seriously considering to dive in and take care of it. I have more time to address stability issues in Triton.

**Comment by blefaudeux** [comment]
> Hey! FYI mem-efficient attention is becoming a bigger priority for us. The bug is pretty deep inside of the compiler but I am seriously considering to dive in and take care of it. I have more time to address stability issues in Triton.

I just updated this old branch, I'm getting an IndexError: map::at at compilation time with the latest dev package, let me know when there's something up to check out  @ptillet !

**Pull request closed without merging**